File: library/src/main/java/android/support/v7/preference/XpPreferenceManager.java
package android.support.v7.preference;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.SharedPreferencesCompat;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author Eugen on 6. 12. 2015.
 */
public final class XpPreferenceManager extends PreferenceManager {

    private static final Method METHOD_SET_NO_COMMIT;

    private static final String[] DEFAULT_PACKAGES;

    static {
        Method setNoCommit = null;
        try {
            setNoCommit = PreferenceManager.class.getDeclaredMethod("setNoCommit", boolean.class);
            setNoCommit.setAccessible(true);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }
        METHOD_SET_NO_COMMIT = setNoCommit;

        if (Build.VERSION.SDK_INT < 14) {
            DEFAULT_PACKAGES = new String[]{"net.xpece.android.support.preference.", "android.support.v7.preference."};
        } else {
            DEFAULT_PACKAGES = new String[]{"net.xpece.android.support.preference.", "android.support.v14.preference.", "android.support.v7.preference."};
        }
    }

    private String[] mCustomDefaultPackages;
    private String[] mAllDefaultPackages;

    private void setNoCommit(boolean noCommit) {
        try {
            METHOD_SET_NO_COMMIT.invoke(this, noCommit);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * @hide
     */
    public XpPreferenceManager(@NonNull final Context context, @Nullable final String[] customDefaultPackages) {
        this(context);
        mCustomDefaultPackages = customDefaultPackages;
    }

    /**
     * @hide
     */
    public XpPreferenceManager(@NonNull final Context context) {
        super(context);
    }

    @Override
    public PreferenceScreen inflateFromResource(@NonNull Context context, int resId, @Nullable PreferenceScreen rootPreferences) {
        this.setNoCommit(true);
        PreferenceInflater inflater = new PreferenceInflater(context, this);
        initPreferenceInflater(inflater);
        rootPreferences = (PreferenceScreen) inflater.inflate(resId, rootPreferences);
        rootPreferences.onAttachedToHierarchy(this);
        this.setNoCommit(false);
        return rootPreferences;
    }

    private void initPreferenceInflater(PreferenceInflater inflater) {
        if (mAllDefaultPackages == null) {
            if (mCustomDefaultPackages == null || mCustomDefaultPackages.length == 0) {
                mAllDefaultPackages = DEFAULT_PACKAGES;
            } else {
                List<String> allDefaultPackagesSet = new ArrayList<>(mCustomDefaultPackages.length + DEFAULT_PACKAGES.length);
                Collections.addAll(allDefaultPackagesSet, mCustomDefaultPackages);
                Collections.addAll(allDefaultPackagesSet, DEFAULT_PACKAGES);
                mAllDefaultPackages = allDefaultPackagesSet.toArray(new String[allDefaultPackagesSet.size()]);
            }
        }
        inflater.setDefaultPackages(mAllDefaultPackages);
    }

    public static void setDefaultValues(@NonNull Context context, int resId, boolean readAgain, @Nullable final String[] customDefaultPackages) {
        setDefaultValues(context, getDefaultSharedPreferencesName(context), getDefaultSharedPreferencesMode(), resId, readAgain, customDefaultPackages);
    }

    public static void setDefaultValues(@NonNull Context context, int resId, boolean readAgain) {
        setDefaultValues(context, getDefaultSharedPreferencesName(context), getDefaultSharedPreferencesMode(), resId, readAgain, null);
    }

    public static void setDefaultValues(@NonNull Context context, @NonNull String sharedPreferencesName, int sharedPreferencesMode, int resId, boolean readAgain) {
        setDefaultValues(context, sharedPreferencesName, sharedPreferencesMode, resId, readAgain, null);
    }

    public static void setDefaultValues(@NonNull Context context, @NonNull String sharedPreferencesName, int sharedPreferencesMode, int resId, boolean readAgain, @Nullable final String[] customDefaultPackages) {
        SharedPreferences defaultValueSp = context.getSharedPreferences(KEY_HAS_SET_DEFAULT_VALUES, 0);
        if (readAgain || !defaultValueSp.getBoolean(KEY_HAS_SET_DEFAULT_VALUES, false)) {
            XpPreferenceManager pm = new XpPreferenceManager(context, customDefaultPackages);
            pm.setSharedPreferencesName(sharedPreferencesName);
            pm.setSharedPreferencesMode(sharedPreferencesMode);
            pm.inflateFromResource(context, resId, null);
            SharedPreferences.Editor editor = defaultValueSp.edit().putBoolean(KEY_HAS_SET_DEFAULT_VALUES, true);
            SharedPreferencesCompat.EditorCompat.getInstance().apply(editor);
        }
    }

    private static String getDefaultSharedPreferencesName(Context context) {
        return context.getPackageName() + "_preferences";
    }

    private static int getDefaultSharedPreferencesMode() {
        return Context.MODE_PRIVATE;
    }
}

File: library/src/main/java/android/support/v7/preference/XpPreferenceManager.kt
package android.support.v7.preference

import android.content.Context
import android.os.Build
import android.support.v4.content.SharedPreferencesCompat
import java.lang.reflect.Method
import java.util.*

/**
 * @author Eugen on 6. 12. 2015.
 */
class XpPreferenceManager(context: Context, val mCustomDefaultPackages: Array<String>? = null) : PreferenceManager(context) {

    private val mAllDefaultPackages: Array<String> by lazy {
        if (mCustomDefaultPackages == null || mCustomDefaultPackages.size == 0) {
            DEFAULT_PACKAGES
        } else {
            val allDefaultPackagesSet = HashSet<String>(mCustomDefaultPackages.size + DEFAULT_PACKAGES.size)
            Collections.addAll(allDefaultPackagesSet, *mCustomDefaultPackages)
            Collections.addAll(allDefaultPackagesSet, *DEFAULT_PACKAGES)
            allDefaultPackagesSet.toTypedArray()
        }
    }

    private fun setNoCommit(noCommit: Boolean) {
        try {
            METHOD_SET_NO_COMMIT!!.invoke(this, noCommit)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun inflateFromResource(context: Context, resId: Int, rootPreferences: PreferenceScreen?): PreferenceScreen {
        this.setNoCommit(true)
        val inflater = PreferenceInflater(context, this)
        inflater.defaultPackages = mAllDefaultPackages
        val newRootPreferences = inflater.inflate(resId, rootPreferences) as PreferenceScreen
        newRootPreferences.onAttachedToHierarchy(this)
        this.setNoCommit(false)
        return newRootPreferences
    }

    companion object {

        private val METHOD_SET_NO_COMMIT: Method?

        private val DEFAULT_PACKAGES: Array<String>

        init {
            var setNoCommit: Method? = null
            try {
                setNoCommit = PreferenceManager::class.java.getDeclaredMethod("setNoCommit", Boolean::class.java)
                setNoCommit!!.isAccessible = true
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }

            METHOD_SET_NO_COMMIT = setNoCommit

            if (Build.VERSION.SDK_INT < 14) {
                DEFAULT_PACKAGES = arrayOf("net.xpece.android.support.preference.", "android.support.v7.preference.")
            } else {
                DEFAULT_PACKAGES = arrayOf("net.xpece.android.support.preference.", "android.support.v14.preference.", "android.support.v7.preference.")
            }
        }

        @JvmStatic
        fun setDefaultValues(context: Context, resId: Int, readAgain: Boolean, customDefaultPackages: Array<String>? = null) {
            setDefaultValues(context, getDefaultSharedPreferencesName(context), getDefaultSharedPreferencesMode(), resId, readAgain, customDefaultPackages)
        }

        @JvmStatic
        fun setDefaultValues(context: Context, sharedPreferencesName: String, sharedPreferencesMode: Int, resId: Int, readAgain: Boolean, customDefaultPackages: Array<String>? = null) {
            val defaultValueSp = context.getSharedPreferences(KEY_HAS_SET_DEFAULT_VALUES, 0)
            if (readAgain || !defaultValueSp.getBoolean(KEY_HAS_SET_DEFAULT_VALUES, false)) {
                val pm = XpPreferenceManager(context, customDefaultPackages)
                pm.sharedPreferencesName = sharedPreferencesName
                pm.sharedPreferencesMode = sharedPreferencesMode
                pm.inflateFromResource(context, resId, null)
                val editor = defaultValueSp.edit().putBoolean(KEY_HAS_SET_DEFAULT_VALUES, true)
                SharedPreferencesCompat.EditorCompat.getInstance().apply(editor)
            }
        }

        @JvmStatic
        private fun getDefaultSharedPreferencesName(context: Context): String =
                context.packageName + "_preferences"

        @JvmStatic
        private fun getDefaultSharedPreferencesMode(): Int =
                Context.MODE_PRIVATE
    }
}
